README.md
  <!-- badges: start -->
  [![Travis build status](https://travis-ci.org/hughparsonage/fy.svg?branch=master)](https://travis-ci.org/hughparsonage/fy)
   [![Codecov test coverage](https://codecov.io/gh/hughparsonage/fy/branch/master/graph/badge.svg)](https://codecov.io/gh/hughparsonage/fy?branch=master)
  <!-- badges: end -->

# fy
Package for financial years


NEWS.md
# fy 0.4.0

## Internal
  - Fix upcoming namespace clash with data.table for `%notin%`

# fy 0.3.0

## Enhancements

* Allow strings like `2013-2014` as a valid financial years. Thanks to @MattCowgill for the suggestion.


# fy 0.2.0

* Added a `NEWS.md` file to track changes to the package.

Compatibility issues:
  * Ditch package grattan's error message 'fy.yr contains non-FYs' since it referred
    to the wrong function argument and was not specific.
  * `na_error` argument now becomes `validate`.

New features:
  * `fy` now supports some dashes in the financial years.


tests/testthat.R
library(testthat)
library(fy)

test_check("fy")


tests/testthat/helper.R
expect_equal <- function(object, expected,
                         check.attributes = FALSE,
                         info = NULL,
                         label = NULL,
                         expected.label = NULL) {
  if (inherits(object, "fy") && is.character(expected)) {
    object <- as.character(object)
  }
  if (inherits(expected, "fy") && is.character(object)) {
    expected <- as.character(expected)
  }
  testthat::expect_equal(object, expected,
                         check.attributes = check.attributes,
                         info = info,
                         label = label,
                         expected.label = expected.label)
}




tests/testthat/test-is_fy.R
context("test-is_fy")

test_that("is_fy works", {
  expect_true(is_fy("2014 15"))
  expect_false(is_fy("2014 14"))
  expect_true(all(is_fy(c("2014-15", "2015-16", "2015-15")) == c(TRUE, TRUE, FALSE)))
})

test_that("Century issue #3", {
  expect_true(is_fy("2014-2015"))
  expect_true(is_fy("1999-2000"))
})


R/utils.R
.getOption <- function (x, default = NULL) {
  ans <- getOption(x)
  if (is.null(ans)) {
    default
  } else {
    ans
  }
}

accel_repetitive_input <- function(x, FUN, ..., THRESHOLD = 2L) {
  .FUN <- match.fun(FUN)
  if (length(x) < THRESHOLD) {
    .FUN(x)
  } else {
    DT <- setDT(list(x = x))
    .subset2(DT[, `:=`("res", .FUN(.BY[[1L]], ...)), by = "x"],
             "res")
  }
}


#' @noRd
#' @param ... Message
#' @param n Level to emit the error message
#' @return Error message but the calling handle at the 'user-level'
stopn <- function(..., n = -sys.nframe()) {
  error_message <- paste0(..., collapse = "")
  if (!interactive() || is_testing()) {
    stop(error_message, call. = FALSE) # nocov
  }
  condition <- function(subclass, message, call = sys.call(-1),
                        ...) {
    structure(class = c(subclass, "condition"),
              list(message = message,
                   call = call), ...)
  }
  custom_stop <- function(subclass,
                          message,
                          call = sys.call(n - 1L),
                          ...) {
    ER <- condition(c("my_error", "error"),
                    message,
                    call = call,
                    ...)
    stop(ER)
  }
  custom_stop(message = error_message)
}

# nocov start
is_testing <- function() {
  requireNamespace("testthat", quietly = TRUE) &&
    testthat::is_testing()
}

# nocov end




R/methods.R
#' @export
"+.fy" <- function(x, h) {
  if (length(h) == 1L || length(h) == length(x)) {
    if (is.integer(h)) {
      out <- yr2fy(fy2yr(x) + h)
    } else if (is.double(h)) {
      if (!all(h == as.integer(h), na.rm = TRUE)) {
        warning("`h = ", deparse(substitute(h)), " was not integerish.")
        h <- as.integer(h)
      }
      out <- yr2fy(fy2yr(x) + h)
    } else {
      stop("non-numeric to arithmetic operation on fy")
    }
  } else {
    stop("`length(h) = ", length(h), "`, yet the length of `x` is ",
         length(x), "`.")
  }
}

#' @export
"-.fy" <- function(x, h) {
  if (length(h) == 1L || length(h) == length(x)) {
    if (is.integer(h)) {
      out <- yr2fy(fy2yr(x) - h)
    } else if (is.double(h)) {
      if (!all(h == as.integer(h), na.rm = TRUE)) {
        warning("`h = ", deparse(substitute(h)), " was not integerish.")
        h <- as.integer(h)
      }
      out <- yr2fy(fy2yr(x) - h)
    } else {
      stop("non-numeric to arithmetic operation on fy")
    }
  } else {
    stop("`length(h) = ", length(h), "`, yet the length of `x` is ",
         length(x), "`.")
  }
}

#' @export
range.fy <- function(...) {
  if (.getOption("verbose", FALSE)) {
    cat("Using range.fy method\n")
  }
  R <- range_fy2yr(..1)
  yr2fy(R)
}

#' @export
min.fy <- function(...) {
  range.fy(...)[1]
}

#' @export
max.fy <- function(...) {
  range.fy(...)[2]
}

# nocov start

#' @export
print.fy <- function(x, ...) {
  if (length(x) < 10) {
    cat("fy (", length(x), "): ", paste0(x, collapse = "  "), sep = "")
  } else {
    cat("fy (", prettyNum(length(x), big.mark = " "), "): ",
        paste(head(x), collapse = "  "), "...\t",
        if (!is.null(miny <- attr(x, "fy_min_yr"))) paste0("  min_yr = ", miny),
        if (!is.null(maxy <- attr(x, "fy_max_yr"))) paste0("  max_yr = ", maxy),
        sep = "")
  }
}
# nocov end




data-raw/put-data.R
library(fastmatch)
fys1901 <- fy::yr2fy(1901:2100)
fys1901B <- sub("-", "", fys1901, fixed = TRUE)
fys1901C <- sub("-", " ", fys1901, fixed = TRUE)
fys1901_2012 <- sub("-", "\u2012", fys1901, fixed = TRUE)
fys1901_2011 <- sub("-", "\u2011", fys1901, fixed = TRUE)
fys1901_2013 <- sub("-", "\u2013", fys1901, fixed = TRUE)
fys1901_2014 <- sub("-", "\u2014", fys1901, fixed = TRUE)

fys1901_9 <- sprintf("%d-%d", 1900:2099, 1901:2100)
fys1901B_9 <- sub("-", "", fys1901_9, fixed = TRUE)
fys1901C_9 <- sub("-", " ", fys1901_9, fixed = TRUE)
fys1901_2012_9 <- sub("-", "\u2012", fys1901_9, fixed = TRUE)
fys1901_2011_9 <- sub("-", "\u2011", fys1901_9, fixed = TRUE)
fys1901_2013_9 <- sub("-", "\u2013", fys1901_9, fixed = TRUE)
fys1901_2014_9 <- sub("-", "\u2014", fys1901_9, fixed = TRUE)


# Generate match hash
"1999-00" %fin% fys1901
"1999-00" %fin% fys1901B
"1999-00" %fin% fys1901C
"1999-00" %fin% fys1901_2012
"1999-00" %fin% fys1901_2011
"1999-00" %fin% fys1901_2013
"1999-00" %fin% fys1901_2014

"1999-2000" %fin% fys1901_9
"1999-2000" %fin% fys1901B_9
"1999-2000" %fin% fys1901C_9
"1999-2000" %fin% fys1901_2012_9
"1999-2000" %fin% fys1901_2011_9
"1999-2000" %fin% fys1901_2013_9
"1999-2000" %fin% fys1901_2014_9

fys1901_all <- c(fys1901, fys1901B, fys1901C)
"1999-00" %fin% fys1901_all


usethis::use_data(fys1901,
                  fys1901B,
                  fys1901C,
                  fys1901_2011,
                  fys1901_2012,
                  fys1901_2013,
                  fys1901_2014,

                  fys1901_9,
                  fys1901B_9,
                  fys1901C_9,
                  fys1901_2011_9,
                  fys1901_2012_9,
                  fys1901_2013_9,
                  fys1901_2014_9,

                  fys1901_all,
                  internal = TRUE, overwrite = TRUE)


R/validate_fys_permitted.R
#' Verifying validity of financial years
#'
#' @description Many functions expect financial years.
#' Determining that they are validly entered is often quite
#' computationally costly, relative to the core calculations.
#' These internal functions provide mechanisms to check validity
#' quickly, while still providing clear, accurate error messages.
#'
#' @aliases valid-fys
#'
#' @param to_verify A user-provided value, purporting to be
#' character vector of financial years.
#' @param permitted_fys A character vector of valid financial years.
#' @param min.yr,max.yr Integers specifying the range of \code{to_verify}.
#' If \code{NULL}, no restriction on the upper or lower bound of the range.
#'
#' @param deparsed A string indicating the argument that the user provided.
#' Should generally be provided explicitly as the default is unlikely
#' to be user-friendly.
#' @param allow.projection If \code{FALSE} emit a different error message.
#' @param earliest_permitted_financial_year,latest_permitted_financial_year Text
#'  for earliest/latest permitted financial year when \code{min.yr}/\code{max.yr}
#'  condition is violated.
#'
#' @param .retain_fmatches If \code{TRUE}, the function may retain an attribute
#' \code{fy_fmatches}  an integer vector of the matches against the financial years
#' \code{"1900-01"} to \code{"2099-00"}. A trade-off between memory and runtime from
#' not recalculating matches.
#'
#' @details The preferred form is \code{"2012-13"}, and this function
#'  returns all elements of \code{to_verify} in this form. That is, it does
#'  not preserve the input form.
#'
#'  Other forms that are recognized (and converted) are:
#'  \itemize{
#'    \item{"201213"}
#'    \item{"2012 13"}
#'    \item{"2012\code{\\u2011}13"}
#'    \item{"2012\code{\\u2012}13"}
#'    \item{"2012\code{\\u2013}13"}
#'    \item{"2012\code{\\u2014}13"}
#'    \item{"2012-2013"}
#'    }
#'
#'
#'
#'
#' @return If \code{to_verify} contains valid financial years
#' they are returned all in the form \code{2013-14}. If they were
#' already in that form, they obtain the following attributes:
#' \describe{
#' \item{\code{fy_all_fy}}{\code{TRUE} if all the financial years are valid.}
#' \item{\code{fy_min_yr}}{An integer, the earliest year ending in \code{to_verify}.}
#' \item{\code{fy_max_yr}}{An integer, the latest year ending in \code{to_verify}.}
#' \item{\code{fy_fmatches}}{An integer vector, the matches with the prebuilt financial years.}
#' }
#'
#' @section Benchmarks:
#'
#' \preformatted{
#'  x <- rep_len(yr2fy(2004L), 1e9)
#'  bench::system_time(validate_fys_permitted(x))
#'  #> process    real
#'  #>  3.578s  3.576s
#'  x <- rep_len(yr2fy(1980:2016), 1e9)
#'
#'  bench::system_time(validate_fys_permitted(x))
#'  #> process    real
#'  #>  3.766s  3.753s
#' }
#'
#'
#' @export validate_fys_permitted



validate_fys_permitted <- function(to_verify,
                                   permitted_fys = NULL,
                                   min.yr = NULL, max.yr = NULL,
                                   deparsed = deparse(substitute(to_verify)),
                                   allow.projection = TRUE,
                                   earliest_permitted_financial_year = "earliest permitted financial year",
                                   latest_permitted_financial_year = "latest permitted financial year",
                                   .retain_fmatches = FALSE) {

  if (!is.character(to_verify)) {
    stopn("`", deparsed, "` was type ", typeof(to_verify), ", ",
          "but must be type character. ",
          "Ensure `", deparsed, "` is a character vector of financial years",
          if (!is.null(min.yr) || !is.null(max.yr)) " satisfying ",
          if (!is.null(min.yr)) paste0("`", yr2fy(min.yr), " <= "),
          if (!is.null(min.yr) || !is.null(max.yr)) deparsed,
          if (!is.null(max.yr)) paste0(" <= ", yr2fy(max.yr), "`"),
          ".")
  }

  if (isTRUE(attr(to_verify, "fy_all_fy"))) {
    # If min.yr and max.yr are fine, we're done
    if (is.null(min.yr) && is.null(max.yr)) {
      return(to_verify)
    }

    # Otherwise we just have to check the ranges: either the ranges
    # are no good (in which case error), or return to_verify

    # min
    if (!is.null(min.yr)) {

      # Unlikely (misspecified), but should assert
      if (is.null(attr(to_verify, "fy_min_yr"))) {
        min_to_verify_yr <- min_fy2yr(to_verify)
        attr(to_verify, "fy_min_yr") <- min_to_verify_yr
      }

      if (min.yr > attr(to_verify, "fy_min_yr")) {
        min.k <- min.yr - 1900L
        stopn("`", deparsed,
              if (length(to_verify) == 1L) " = " else "` contained ",
              '"', fys1901[attr(to_verify, "fy_min_yr") - 1900L], '"',
              if (length(to_verify) == 1L) "`",
              " which ",
              "is earlier than the ",
              earliest_permitted_financial_year,
              ": ", '"', fys1901[min.k], '"', ".")
      }
    }

    # max
    if (!is.null(max.yr)) {

      # Unlikely (misspecified), but should assert
      if (is.null(attr(to_verify, "fy_max_yr"))) {
        max_to_verify_yr <- max_fy2yr(to_verify)
        attr(to_verify, "fy_max_yr") <- max_to_verify_yr
      }

      if (max.yr < attr(to_verify, "fy_max_yr")) {
        max.k <- max.yr - 1900L
        deparsed <- force(deparsed)
        stopn(if (isFALSE(allow.projection)) "`allow.projection = FALSE`, yet ",
              "`", deparsed,
              if (length(to_verify) == 1L) " = " else "` contained ",
              '"', fys1901[attr(to_verify, "fy_max_yr") - 1900L], '"',
              if (length(to_verify) == 1L) "`",
              " which ",
              "is later than the ",
              latest_permitted_financial_year,
              ": ", '"', fys1901[max.k], '"', ".")
      }
    }

    return(to_verify)
  }



  if (is.null(permitted_fys)) {
    # Use the anyNA virtue of hutils::coalesce
    if (anyNA3i(fmatches <- fmatch(to_verify, fys1901))) {
      # Not standard financial year like '2014-15'.
      # Is it just a nonstandard but valid financial year
      # or is it not a financial year at all?
      fmatches <- coalesce3i(fmatches,
                             fmatch(to_verify, fys1901B),
                             fmatch(to_verify, fys1901C),
                             fmatch(to_verify, fys1901_2011),
                             fmatch(to_verify, fys1901_2012),
                             fmatch(to_verify, fys1901_2013),
                             fmatch(to_verify, fys1901_2014),
                             fmatch(to_verify, fys1901_9),
                             fmatch(to_verify, fys1901B_9),
                             fmatch(to_verify, fys1901C_9),
                             fmatch(to_verify, fys1901_2011_9),
                             fmatch(to_verify, fys1901_2012_9),
                             fmatch(to_verify, fys1901_2013_9),
                             fmatch(to_verify, fys1901_2014_9))
      if (anyNA3i(fmatches)) {
        first_bad <- which.max(is.na(fmatches))
        stopn("`", deparsed,
              if (length(to_verify) == 1L) " = " else "` contained ",
              '"', to_verify[first_bad], '"',
              if (length(to_verify) == 1L) "` was " else " which is ",
              "not a valid financial year.")
      }

      # Standardize
      to_verify <- fys1901[fmatches]
    }

    class(to_verify) <- "fy"
    attr(to_verify, "fy_all_fy") <- TRUE
    if (!is.null(min.yr)) {
      min.k <- min.yr - 1900L
      min_fmatches <- min(fmatches)
      if (min_fmatches < min.k) {
        first_bad <- which.min(fmatches)
        stopn("`", deparsed,
              if (length(to_verify) == 1L) " = " else "` contained ",
              '"', to_verify[first_bad], '"',
              if (length(to_verify) == 1L) "`",
              " which ",
              "is earlier than the ",
              earliest_permitted_financial_year,
              ": ", '"', fys1901[min.k], '"', ".")
      }
      attr(to_verify, "fy_min_yr") <- min_fmatches + 1900L
    }
    if (!is.null(max.yr)) {
      max.k <- max.yr - 1900L
      max_fmatches <- max(fmatches)
      if (max_fmatches > max.k) {
        first_bad <- which.max(fmatches)
        stopn(if (isFALSE(allow.projection)) "`allow.projection = FALSE`, yet ",
              "`", deparsed,
              if (length(to_verify) == 1L) " = " else "` contained ",
              '"', to_verify[first_bad], '"',
              if (length(to_verify) == 1L) "`",
              " which ",
              "is later than the ",
              latest_permitted_financial_year,
              ": ", '"', fys1901[max.k], '"', ".")
      }
      attr(to_verify, "fy_max_yr") <- max_fmatches + 1900L
    }

    if (isTRUE(.retain_fmatches)) {
      attr(to_verify, "fy_fmatches") <- fmatches
    }

    return(invisible(to_verify))

  } else {
    permitted_fys <- validate_fys_permitted(permitted_fys)
  }


  fy.year <- to_verify
  if (!all(fy.year %chin% permitted_fys)) {
    if (any(!is_fy(fy.year))) {
      i <- which(!is_fy(fy.year))
      i1 <- i[1]
      if (length(i) > 1) {
        stopn("`", deparsed, "` contained invalid FYs. ",
              "There were ",
              length(i), " invalid entries (",
              round(100 * length(i) / length(fy.year)), "%).",
              "\n\n",
              "First invalid FY:\n\t", fy.year[i1], "\n",
              "at position ", i)
      } else {
        if (length(fy.year) == 1L) {
          stopn("`", deparsed, "` set to '", fy.year, "', was not a valid financial year. ",
                "Select a valid fy.year between ",
                permitted_fys[1], " and ", last(permitted_fys), ".")
        } else {
          stopn("`", deparsed, "` contained invalid entry ",
                fy.year[i1], " at position ", i1, ".")
        }
      }
    } else {
      # all are valid
      if (min.fy(fy.year) < min.fy(permitted_fys) ||
          max.fy(fy.year) > max.fy(permitted_fys)) {

        i <- which(fy.year %notin% permitted_fys)
        i1 <- i[1]

        if (length(i) == 1L) {
          stopn("`", deparsed, " = ", fy.year[i1], "` was not within the allowed range: ",
                permitted_fys[1], " <= fy.year <= ", last(permitted_fys))
        } else {
          stopn("`", deparsed, "` were not within the allowed range: ",
                permitted_fys[1], " <= fy.year <= ", last(permitted_fys), "\n\n",
                "First invalid FY:\n\t", fy.year[i1], "\n",
                "at position ", i1)
        }
      }
    }
  }
  class(to_verify) <- "fy"
  return(to_verify)
}











R/fy_year.R
#' Convenience functions for dealing with financial years
#'
#' @name is_fy
#' @aliases fy.year yr2fy fy2yr fy2date date2fy
#' @param yr_ending An integer representing a year.
#' @param x A character vector suspected to be a financial year.
#' @param yq A character vector representing year quarters in \code{1066-Q2} format.
#' @param date A string or date for which the financial year is desired. Note that \code{yr2fy} does not check its argument is an integer.
#' @param validate If \code{TRUE}, the default, inputs that are expected to be financial years
#' are first validated. Validation should be very fast, though some use-cases may require this be skipped.
#' @param assume1901_2100 For \code{yr2fy}, assume that \code{yr_ending} is between 1901 and 2100,
#' for performance. By default, set to \code{getOption("fy.assume1901_2100", TRUE)}.
#'
#'
#' @details See \code{\link{valid-fys}} for allowed forms of \code{x}.
#'
#' @return For \code{is_fy}, a logical, whether its argument is a financial year.
#' The following forms are allowed: \code{2012-13}, \code{201213}, \code{2012 13}, as well as
#' \code{2012<dash>13} for some dash symbols.
#' For \code{fy.year}, \code{yr2fy}, and \code{date2fy}, the financial year.
#' For the inverses, a numeric corresponding to the year.
#'
#' \code{fy.year} was an alias for \code{yr2fy}, and is now defunct.
#'
#' \code{fy2yr} converts a financial year to the year ending: \code{fy2yr("2016-17")} returns 2017.
#'  \code{yr2fy} is the inverse: \code{yr2fy(fy2yr("2016-17")) == "2016-17"}.
#'
#' \code{fy2date} converts a financial year to the 30 June of the financial year ending.
#'
#' \code{date2fy} converts a date to the corresponding financial year.
#'
#'
#' @examples
#' is_fy("2012-13")
#' is_fy("2012-14")
#' yr2fy(2012)
#' fy2yr("2015-16")
#' date2fy("2014-08-09")
#' @export is_fy
NULL

is_fy <- function(x) {
  if (length(x) > 1L) {
    return(accel_repetitive_input(x, is_fy, THRESHOLD = 2L))
  }
  if (anyNA(x)) {
    return(NA)
  }
  if (!is.character(x)) {
    return(FALSE)
  }
  # strsplit(x, split = "(?<=[0-9]{4})(?=[0-9])|(?<=[0-9]{4})(?=[[:punct:]])|(?<=[[:punct:]])(?=[0-9])", perl = T)
  #
  # strsplit(split)

  fy_split <- strsplit(x, "((?<=([0-9]{4}))[^0-9]?)", perl = TRUE)
  yr0 <- fy_split[[1]][1]
  yr1 <- fy_split[[1]][2]

  yearlike <- grepl("^[0-9][0-9]+$", yr0) && grepl("^[0-9][0-9](?:[0-9]{2})?$", yr1)
  if (!yearlike) {
    return(FALSE)
  }

  if (nchar(yr1) == 2L) {
    yr1_from_lhs <- (as.integer(yr0) + 1L) %% 100L
    yr1_from_rhs <- as.integer(yr1) %% 100L
  }  else {
    # 2013, 14
    yr1_from_lhs <- as.integer(yr0) + 1L
    yr1_from_rhs <- as.integer(yr1)
  }

  if (yr1_from_lhs != yr1_from_rhs) {
    return(FALSE)
  }

  ncharx <- nchar(x)
  if (ncharx == 7L || ncharx == 9L) {
    sep <- substr(x, 5L, 5L)
    if (sep != "-" &&
        sep != " " &&
        sep != "\u2011" &&
        sep != "\u2012" &&
        sep != "\u2013" &&
        sep != "\u2014") {
      return(FALSE)
    }
  }

  return(TRUE)
}

is_fy2 <- function(x) {
  a <- c("1897-98", "1898-99", "1899-00", "1900-01", "1901-02", "1902-03",
         "1903-04", "1904-05", "1905-06", "1906-07", "1907-08", "1908-09",
         "1909-10", "1910-11", "1911-12", "1912-13", "1913-14", "1914-15",
         "1915-16", "1916-17", "1917-18", "1918-19", "1919-20", "1920-21",
         "1921-22", "1922-23", "1923-24", "1924-25", "1925-26", "1926-27",
         "1927-28", "1928-29", "1929-30", "1930-31", "1931-32", "1932-33",
         "1933-34", "1934-35", "1935-36", "1936-37", "1937-38", "1938-39",
         "1939-40", "1940-41", "1941-42", "1942-43", "1943-44", "1944-45",
         "1945-46", "1946-47", "1947-48", "1948-49", "1949-50", "1950-51",
         "1951-52", "1952-53", "1953-54", "1954-55", "1955-56", "1956-57",
         "1957-58", "1958-59", "1959-60", "1960-61", "1961-62", "1962-63",
         "1963-64", "1964-65", "1965-66", "1966-67", "1967-68", "1968-69",
         "1969-70", "1970-71", "1971-72", "1972-73", "1973-74", "1974-75",
         "1975-76", "1976-77", "1977-78", "1978-79", "1979-80", "1980-81",
         "1981-82", "1982-83", "1983-84", "1984-85", "1985-86", "1986-87",
         "1987-88", "1988-89", "1989-90", "1990-91", "1991-92", "1992-93",
         "1993-94", "1994-95", "1995-96", "1996-97", "1997-98", "1998-99",
         "1999-00", "2000-01", "2001-02", "2002-03", "2003-04", "2004-05",
         "2005-06", "2006-07", "2007-08", "2008-09", "2009-10", "2010-11",
         "2011-12", "2012-13", "2013-14", "2014-15", "2015-16", "2016-17",
         "2017-18", "2018-19", "2019-20", "2020-21", "2021-22", "2022-23",
         "2023-24", "2024-25", "2025-26", "2026-27", "2027-28", "2028-29",
         "2029-30", "2030-31", "2031-32", "2032-33", "2033-34", "2034-35",
         "2035-36", "2036-37", "2037-38", "2038-39", "2039-40", "2040-41",
         "2041-42", "2042-43", "2043-44", "2044-45", "2045-46", "2046-47",
         "2047-48", "2048-49", "2049-50", "2050-51", "2051-52", "2052-53")
  x %fin% a
}

all_fy <- function(x, permitted = NULL) {
  is.character(x) && length(x) && {


    a <- if (is.null(permitted)) {
      c("1897-98", "1898-99", "1899-00", "1900-01", "1901-02", "1902-03",
        "1903-04", "1904-05", "1905-06", "1906-07", "1907-08", "1908-09",
        "1909-10", "1910-11", "1911-12", "1912-13", "1913-14", "1914-15",
        "1915-16", "1916-17", "1917-18", "1918-19", "1919-20", "1920-21",
        "1921-22", "1922-23", "1923-24", "1924-25", "1925-26", "1926-27",
        "1927-28", "1928-29", "1929-30", "1930-31", "1931-32", "1932-33",
        "1933-34", "1934-35", "1935-36", "1936-37", "1937-38", "1938-39",
        "1939-40", "1940-41", "1941-42", "1942-43", "1943-44", "1944-45",
        "1945-46", "1946-47", "1947-48", "1948-49", "1949-50", "1950-51",
        "1951-52", "1952-53", "1953-54", "1954-55", "1955-56", "1956-57",
        "1957-58", "1958-59", "1959-60", "1960-61", "1961-62", "1962-63",
        "1963-64", "1964-65", "1965-66", "1966-67", "1967-68", "1968-69",
        "1969-70", "1970-71", "1971-72", "1972-73", "1973-74", "1974-75",
        "1975-76", "1976-77", "1977-78", "1978-79", "1979-80", "1980-81",
        "1981-82", "1982-83", "1983-84", "1984-85", "1985-86", "1986-87",
        "1987-88", "1988-89", "1989-90", "1990-91", "1991-92", "1992-93",
        "1993-94", "1994-95", "1995-96", "1996-97", "1997-98", "1998-99",
        "1999-00", "2000-01", "2001-02", "2002-03", "2003-04", "2004-05",
        "2005-06", "2006-07", "2007-08", "2008-09", "2009-10", "2010-11",
        "2011-12", "2012-13", "2013-14", "2014-15", "2015-16", "2016-17",
        "2017-18", "2018-19", "2019-20", "2020-21", "2021-22", "2022-23",
        "2023-24", "2024-25", "2025-26", "2026-27", "2027-28", "2028-29",
        "2029-30", "2030-31", "2031-32", "2032-33", "2033-34", "2034-35",
        "2035-36", "2036-37", "2037-38", "2038-39", "2039-40", "2040-41",
        "2041-42", "2042-43", "2043-44", "2044-45", "2045-46", "2046-47",
        "2047-48", "2048-49", "2049-50", "2050-51", "2051-52", "2052-53")
    } else {
      permitted
    }
    if (anyNA(fmatch(x, a))) {
      a <- NULL
      FALSE
    } else {
      a <- NULL
      TRUE
    }
  }
}

range_fy2yr <- function(x) {
  # Put before any fmatch
  if (!is.atomic(x)) {
    stop("`x` was class ", toString(class(x)),
         " but must be atomic.")
  }

  if (length(x) == 1L) {
    y <- fmatch(x, fys1901) + 1900L
    return(rep(y, times = 2L))
  }
  if (!is.null(g_min_yr <- attr(x, "fy_min_yr")) &&
      !is.null(g_max_yr <- attr(x, "fy_max_yr"))) {
    return(c(g_min_yr, g_max_yr))
  }


  y <- fmatch(x, fys1901)

  miny <- min(y, na.rm = TRUE) + 1900L
  maxy <- max(y, na.rm = TRUE) + 1900L

  if (is.symbol(substitute(x))) {
    setattr(x, "fy_min_yr", miny)
    setattr(x, "fy_max_yr", maxy)
  }
  c(miny, maxy)
}

min_fy2yr <- function(x) {
  range_fy2yr(x)[1L]
}

max_fy2yr <- function(x) {
  range_fy2yr(x)[2L]
}

#' @rdname is_fy
#' @export yr2fy
yr2fy <- function(yr_ending,
                  assume1901_2100 = .getOption("fy.assume1901_2100",
                                               .getOption("grattan.assume1901_2100",
                                                          TRUE))) {
  if (!is.atomic(yr_ending) || !is.numeric(yr_ending)) {
    stop("`yr_ending` was class ", toString(class(yr_ending)),
         " but must be an atomic, numeric vector.")
  }

  out <-
    if (assume1901_2100 ||
        AND(min(yr_ending) > 1900L,
            max(yr_ending) < 2100L)) {
      fys1901[yr_ending - 1900L]
    } else {
      .yr2fy(yr_ending)
    }
  class(out) <- "fy"
  out
}


#' @rdname is_fy
#' @export .yr2fy
.yr2fy <- function(yr_ending) {
  if (!is.atomic(yr_ending) || !is.numeric(yr_ending)) {
    stop("`yr_ending` was class ", toString(class(yr_ending)),
         " but must be atomic.")
  }
  out <-
    if (length(yr_ending) > 10e3L) {
      # Apparently quicker for > 1000
      accel_repetitive_input(yr_ending, .yr2fy)
    } else {
      sprintf("%d-%02d", as.integer(yr_ending) - 1L, as.integer(yr_ending) %% 100L)
    }
  class(out) <- "fy"
  out
}

#' @rdname is_fy
#' @export fy2yr
fy2yr <- function(x, validate = TRUE) {

  if (isTRUE(validate)) {
    x <- validate_fys_permitted(x)
  } else if (!is.atomic(x)) {
    # Need to check this even if validate isn't
    # TRUE, because `fmatch` will complain loudly.
    stop("`x` was class ", toString(class(x)),
         " but must be atomic.")
  }

  1900L + fmatch(x, fys1901)
}


#' @rdname is_fy
#' @export fy2date
fy2date <- function(x, validate = TRUE) {
  if (isTRUE(validate)) {
    x <- validate_fys_permitted(x)
  }
  if (length(x) == 1L) {
    date <- paste0(as.integer(substr(x, 0L, 4L)) + 1L, "-06-30")
    return(as.Date(date))
  }
  accel_repetitive_input(x, fy2date, validate = FALSE)
}


#' @rdname is_fy
#' @export date2fy
date2fy <- function(date) {
  if (!inherits(date, "Date")) {
    date <- as.Date(date)
  }
  out <- yr2fy(year(date) + {month(date) >= 7L})
  class(out) <- "fy"
  out
}


#' @rdname is_fy
#' @export qtr2fy
qtr2fy <- function(yq) {
  if (inherits(yq, "yearqtr")) {
    yqn <- as.numeric(yq)
    o <-
      yr2fy(if_else(yqn %% 1 >= 0.5,
                    yqn + 1,
                    yqn))
    o
  } else if (is.character(yq)) {
    # Rely on the first element to determine the
    # format
    first_yq <- yq[1L]
    if (is.na(first_yq)) {
      yq_is_na <- is.na(yq)
      first_yq <- first(yq[which.min(yq_is_na)])
    }

    y <- q <- NULL
    cm <- CJ(y = 1901:2099, q = 1:4)
    cm[, "YQ" := sprintf("%d%sQ%d", y, substr(first_yq, 5L, 5L), q)]
    cm[, "fy_year" := yr2fy(y + q %in% 3:4)]
    cmyq <- .subset2(cm, "YQ")
    o <- .subset2(cm, "fy_year")[fmatch(yq, cmyq)]

  } else {
    stop("Unknown class for `yq`.")
  }
  class(o) <- "fy"
  o
}

